yarn/src/main/scala/org/apache/spark/deploy/yarn/ContainerPlacementStrategy.scala

```diff
@@ -0,0 +1,196 @@
+      hostToLocalTaskCount: Map[String, Int]
+    ): Array[ContainerLocalityPreferences]
+}
+
+/**
+ * This strategy is calculating the optimal locality preferences of YARN containers by considering
+ * the node ratio of pending tasks, number of required cores/containers and and locality of current
+ * existed containers. The target of this algorithm is to maximize the number of tasks that
+ * would run locally.
+ *
+ * The details of this algorithm is described as below, if we have 20 tasks which
+ * require (host1, host2, host3) and 10 tasks which require (host1, host2, host4),
+ * besides each container has 2 cores and cpus per task is 1, so the required container number is
+ * 15, and host ratio is (host1: 30, host2: 30, host3: 20, host4: 10).
+ *
+ * 1. If requested container number (18) is more than the required container number (15):
+ *
+ * requests for 5 containers with nodes: (host1, host2, host3, host4)
+ * requests for 5 containers with nodes: (host1, host2, host3)
+ * requests for 5 containers with nodes: (host1, host2)
+ * requests for 3 containers with no locality preferences.
+ *
+ * The placement ratio is 3 : 3 : 2 : 1, and set the additional containers with no locality
+ * preferences.
+ *
+ * 2. If requested container number (10) is less than or equal to the required container number
+ * (15):
+ *
+ * requests for 4 containers with nodes: (host1, host2, host3, host4)
+ * requests for 3 containers with nodes: (host1, host2, host3)
+ * requests for 3 containers with nodes: (host1, host2)
+ *
+ * The placement ratio is 10 : 10 : 7 : 4, close to expected ratio (3 : 3 : 2 : 1)
+ *
+ * 3. If containers are existed but no matching localities, follow the method of 1 and 2.
+ *
+ * 4. If containers are existed and some localities are matched. For example if we have 1
+ * containers on each node (host1: 1, host2: 1: host3: 1, host4: 1), and the expected containers
+ * on each node would be (host1: 5, host2: 5, host3: 4, host4: 2),
+ * so the newly requested containers on each node would be updated to (host1: 4, host2: 4,
+ * host3: 3, host4: 1), 12 containers by total.
+ *
+ *   4.1 If requested container number (18) is more than newly required containers (12). Follow
+ *   method 1 with updated ratio 4 : 4 : 3 : 1.
+ *
+ *   4.2 If request container number (10) is more than newly required containers (12). Follow
+ *   method 2 with updated ratio 4 : 4 : 3 : 1.
+ *
+ * 5. If containers are existed and existing localities can fully cover the requested localities.
+ * For example if we have 5 containers on each node (host1: 5, host2: 5, host3: 5, host4: 5),
+ * which could cover the current requested localities. This algorithm will allocate all the
+ * requested containers with no localities.
+ */
+private[yarn] class LocalityPreferredContainerPlacementStrategy(
+    val sparkConf: SparkConf,
+    val yarnConf: Configuration,
+    val yarnAllocator: YarnAllocator) extends ContainerPlacementStrategy {
```

Comment:
Instead of passing in the whole yarnAllocator, can we pass in `resource` and `allocatedHostToContainersMap` explicitly as arguments to the `localityOfRequestedContainers` method so that it's clear what information is used to make our decision?